javascript/modern-js.md

```diff
@@ -190,69 +206,153 @@ showResult(name, subject, grade);
-  return parsedData;
+async function fetchData(url) {
+  const response = await fetch(url);
+  if (response.ok) {
+    return response.json();
+  }
+  throw new Error(`HTTP Error ${response.status}:  ${response.statusText}`);
 }
 ```
 
-How do we use it? We put the keyword `async` in front of the function declaration that will contain asynchronous code. Then in every line that returns a Promise we put the keyword `await` in front. That's it. Now the function will wait for the return of the fetch promise and grabbing the data before continuing to the next line.
+Putting the keyword `async` in front of the function declaration will make this an asynchronous function. This has two effects:
+
+1. This function will now return a promise. This promise will resolve to whatever value you `return`.
+2. You can now use the `await` keyword inside the function body to "await" the settlement of a promise. If the promise is resolved the resolved value becomes available for subsequent use. In this example it is assigned to a variable. If the promise is rejected the `await` keyword effectively [throws](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw) an error (i.e. the rejection value). Note that waiting for the promise to be settled is non-blocking.
+3. Conversely, if you `throw` an error inside an `async` function it will return a rejected promise.
 
 ### Catching errors in async/await
-As you might have noticed, the `async/await` keywords don't give us a way catching errors like it does in the Promise object.
 
-But before we get into that, we should define "catching errors" a little bit. By "catching errors" we mean:
+As you might have noticed, the `async/await` keywords don't give us a way to catch errors like it does in the Promise object.
+
+But before we get into that, we should discuss the type of errors that can typically occur in a JavaScript application.
+
+- Your application can cause runtime errors because of incorrect JavaScript syntax or an incorrect data type. These are typically errors that you need to fix before shipping your application. Your application should not try and catch such errors. They will usually turn up as unhandled errors in the console. Make fixing them your top priority.
 
-1. that a line of code has caused an error (because of incorrect syntax or data type)
-2. that the program has shutdown to prevent any other errors from happening
-3. that the application gives feedback to the developer and/or user about where the error came from
+- Anticipated application and/or infrastructure errors. Examples are network errors, data corruption errors, data validation errors, etc. These types of errors can be expected and should not crash your application if they occur.Instead your application should be resilient to them and recover as gracefully as possible. This might involve informing the user of the situation and suggesting remedial action where possible.
```

Comment:
Would it be interesting to insist that defining expected unhappy path as well as happy path is important when writing code? Could also note that it will help later when testing all the possible behaviors of the program.

Reply:
In the repo that I use for Q&A I included this text:

> We generally talk about a _happy path_ and an _error path_ (or _unhappy path_) for your code.
>
> - The _happy path_ is the execution path through your code if all is well: no errors encountered.
> - The _error path_ is the execution path through your code if an error is encountered.
>
> This section is about designing the _error path_.

Do you mean something like this? If not, can you suggest some text that addresses your point?